### PR TITLE
Adding links to namespaces has undesireable UX

### PR DESCRIPTION
**Pull request by MilanPospisil**
Issue: AAH-149

https://issues.redhat.com/browse/AAH-149

This issue is to track some follow-on work from [AAH-18](https://issues.redhat.com/browse/AAH-18). Now at least adding links works mostly as expected, but there is still some undesirable behavior:

    Links without http:// are valid and users will enter them. "Enter a valid URL" does not make it clear what is missing, either. Schema can be added if missing.
    Every time the namespace is saved, the order of the links is reversed.

### CODE REVIEW

**Comment by MilanPospisil** [comment]
Before:
URLbefore
After:
URLAfter

**Comment by MilanPospisil** [comment]
Note that Save button is still active (the previous functionality stays and button does not save invalid data and displays its own error message). Should it be disabled?

File: src/utilities/validateURLHelper.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { t } from '@lingui/macro';
+
+export function validateURLHelper() {
```
Comment (himdel) [inline src/utilities/validateURLHelper.ts:3]: :+1: on extracting the helper, but why a function returning another function? Can you just export the inner function?

**Comment by himdel** [comment]
> Note that Save button is still active (the previous functionality stays and button does not save invalid data and displays its own error message). Should it be disabled?

Ah, ideally yes, when a link fails UI validation, the Save button should be disabled.

Once UI validation is successful, the user can submit and then possibly get hit with the errors from the API from any other model validations.

But I think this is still missing the part where we take the errors from the server response, and show them by the right field as well. Let me know if you need any help with that, I think the error splitting logic is implemented in `mapErrorMessages`, hopefully it's also compatible with how this form is written :)

**Comment by MilanPospisil** [comment]
> > Note that Save button is still active (the previous functionality stays and button does not save invalid data and displays its own error message). Should it be disabled?
> 
> Ah, ideally yes, when a link fails UI validation, the Save button should be disabled.
> 
> Once UI validation is successful, the user can submit and then possibly get hit with the errors from the API from any other model validations.
> 
> But I think this is still missing the part where we take the errors from the server response, and show them by the right field as well. Let me know if you need any help with that, I think the error splitting logic is implemented in `mapErrorMessages`, hopefully it's also compatible with how this form is written :)

The server does not return errors that could be easily assigned to particular items. For example for 3 invalid items (2 blank names and 3 invalid urls) it says:

0: {status: "400", code: "invalid", title: "Invalid input.", detail: "'asdsadads' is not a valid url.",…}
code: "invalid"
detail: "'asdsadads' is not a valid url."
source: {parameter: "links__url"}
status: "400"
title: "Invalid input."
1: {status: "400", code: "blank", title: "Invalid input.", detail: "This field may not be blank.",…}
code: "blank"
detail: "This field may not be blank."
source: {parameter: "links__name"}
status: "400"
title: "Invalid input."
2: {status: "400", code: "invalid", title: "Invalid input.",…}
code: "invalid"
detail: "'asddasdasb**-ui/pull/1490/files' is not a valid url."
source: {parameter: "links__url"}
status: "400"
title: "Invalid input."
3: {status: "400", code: "blank", title: "Invalid input.", detail: "This field may not be blank.",…}
code: "blank"
detail: "This field may not be blank."
source: {parameter: "links__name"}
status: "400"
title: "Invalid input."
4: {status: "400", code: "invalid", title: "Invalid input.", detail: "'gfhfhf' is not a valid url.",…}
code: "invalid"
detail: "'gfhfhf' is not a valid url."
source: {parameter: "links__url"}
status: "400"
title: "Invalid input."

I suggest that better way is to do good client side validation instead. What do you think?

**Comment by MilanPospisil** [comment]
Another thing. Company name field has IsRequired in the textinput, but no error message is shown when it is empty, nor the server returns any errors. The IsRequired should not be here i guess?

File: src/components/namespace-form/namespace-form.tsx [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import { t } from '@lingui/macro';
 import * as React from 'react';
 import './namespace-form.scss';
+import validateURLHelper from './../../utilities/validateURLHelper';
```
Comment (himdel) [inline src/components/namespace-form/namespace-form.tsx:4]:
```suggestion
import validateURLHelper from 'src/utilities/validateURLHelper';
```

File: src/components/namespace-form/namespace-form.tsx [hunk 3]
```diff
@@ -233,27 +234,76 @@ export class NamespaceForm extends React.Component<IProps, IState> {
...
+    validated: 'default' | 'error';
+    helperTextInvalid?: string;
+  } {
+    if (link.url) {
+      if (link.name) {
+        return { validated: 'default' };
+      } else {
+        return {
+          validated: 'error',
+          helperTextInvalid: 'Name must not be empty.',
```
Comment (himdel) [inline src/components/namespace-form/namespace-form.tsx:247]:
```suggestion
          helperTextInvalid: t`Name must not be empty.`,
```

File: src/components/namespace-form/namespace-form.tsx [hunk 4]
```diff
@@ -233,27 +234,76 @@ export class NamespaceForm extends React.Component<IProps, IState> {
...
+  public static validateUrl(link): ReturnType<typeof validateURLHelper> {
+    if (link.url) {
+      // only validate url if input is not blank, blank inputs are thrown away
+      return validateURLHelper(undefined, link.url);
+    }
+
+    if (link.name) {
+      return {
+        validated: 'error',
+        helperTextInvalid: 'URL must not be empty.',
```
Comment (himdel) [inline src/components/namespace-form/namespace-form.tsx:265]:
```suggestion
        helperTextInvalid: t`URL must not be empty.`,
```

File: src/components/repositories/remote-form.tsx [hunk 5]
```diff
@@ -26,39 +26,7 @@ import { RemoteType, WriteOnlyFieldType } from 'src/api';
...
-  } catch (_) {
-    // fallthrough
-  }
-
-  return {
-    validated: 'error',
-    helperTextInvalid: t`The URL needs to be in 'http(s)://' format.`,
-  };
-};
+import validateURLHelper from './../../utilities/validateURLHelper';
```
Comment (himdel) [inline src/components/repositories/remote-form.tsx:29]:
```suggestion
import validateURLHelper from 'src/utilities/validateURLHelper';
```

File: src/components/namespace-form/namespace-form.tsx [hunk 6]
```diff
@@ -233,27 +234,76 @@ export class NamespaceForm extends React.Component<IProps, IState> {
...
-          <TextInput
-            id='name'
-            type='text'
-            placeholder={t`Link text`}
-            value={link.name}
-            onChange={(value, event) => this.updateLink(index, value, event)}
-          />
+          <FormGroup
+            fieldId={'name'}
+            validated={NamespaceForm.validateName(link).validated}
```
Comment (himdel) [inline src/components/namespace-form/namespace-form.tsx:279]:
```suggestion
```

(the split on the next line already does that)

File: src/components/namespace-form/namespace-form.tsx [hunk 7]
```diff
@@ -233,27 +234,76 @@ export class NamespaceForm extends React.Component<IProps, IState> {
...
-          <TextInput
-            id='url'
-            type='text'
-            placeholder={t`Link URL`}
-            value={link.url}
-            onChange={(value, event) => this.updateLink(index, value, event)}
-          />
+          <FormGroup
+            fieldId={'link'}
+            validated={NamespaceForm.validateUrl(link).validated}
```
Comment (himdel) [inline src/components/namespace-form/namespace-form.tsx:295]:
```suggestion
```

File: src/containers/edit-namespace/edit-namespace.tsx [hunk 8]
```diff
@@ -212,6 +213,20 @@ class EditNamespace extends React.Component<RouteComponentProps, IState> {
...
+    let disabled = false;
+    namespace.links.forEach((link) => {
+      if (
+        NamespaceForm.validateName(link).validated == 'error' ||
+        NamespaceForm.validateUrl(link).validated == 'error'
+      ) {
+        disabled = true;
+      }
+    });
+    return disabled;
```
Comment (himdel) [inline src/containers/edit-namespace/edit-namespace.tsx:227]:
```suggestion
    return namespace.links.some((link) =>
      NamespaceForm.validateName(link).validated == 'error' ||
      NamespaceForm.validateUrl(link).validated == 'error');
```

(+- prettier :))

[Array.some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some)

**Comment by himdel** [comment]
> The server does not return errors that could be easily assigned to particular items...
> I suggest that better way is to do good client side validation instead.

Well we still need to show server errors when they do happen, but agreed we can't really show server errors together with the right item, and clients side validation is enough to make it nice :)

**Comment by himdel** [comment]
> Another thing. Company name field has IsRequired in the textinput, but no error message is shown when it is empty, nor the server returns any errors. The IsRequired should not be here i guess?

Yeah agreed, I'm pretty sure the company name is not required now. There used to be an `isRequired` on the corresponding `FormGroup` before #50 but I'm not sure this was ever required. (And it seems the `TextInput` `isRequired` prop only really sets the `<input required>` prop, which is ignored in `<form novalidate>` (html forms vs react forms).)

So, harmless, but feel free to remove it, I agree it doesn't belong there :)

**Comment by himdel** [comment]
Thanks, the previous issue is fixed, just one more detail :) ...



the width of the right input apparently depends on the presence/length of the message under it, we should keep the 50:50 layout as before

File: src/components/namespace-form/namespace-form.scss [hunk 9]
```diff
@@ -27,18 +27,19 @@
   margin-bottom: 10px;
 
   .link-name {
-    flex-grow: 1;
+    flex: 1;
```
Comment (himdel) [inline src/components/namespace-form/namespace-form.scss:30]: can we keep flex-grow as is, and just add `flex-basis: 0`? (to both)
should have the same effect and makes it more clear what the problem was :)

(Either way LGTM :+1:, fixes the problem :))

**Pull request merged**